src/equal.c

```diff
@@ -1,68 +1,72 @@
 #include "vctrs.h"
 
-int equal_scalar(SEXP x, int i, SEXP y, int j, bool na_equal) {
-  if (TYPEOF(x) != TYPEOF(y))
-    return false;
+int lgl_equal_scalar(int* x, int* y, bool na_equal) {
+  if (*x == NA_LOGICAL) return na_equal ? *y == NA_LOGICAL : NA_LOGICAL;
```

Comment:
I wonder if it's worth rewriting to something like:

```R
if (na_equal) {
  return *x == *y 
} else {
  return (*x == NA_LOGICAL || *y == NA_LOGICAL) ? NA_LOGICAL : *x == *y
}
```

I think that's easier to understand, and you're in a good position to see if it has any performance implications.

Reply:
I have a vague feeling it might be worth re-checking if there's any difference in performance pulling the `na_equal` comparison after this change..... Now I've read further down, that would clearly introduce a bunch of duplication, so it's probably not worth trying for now.

Reply:
I tried this in the `template_full` case (now `template_no_branch` in the new graphic) and it made no difference. I expected to see some difference as well, especially with a high proportion of missing values.